File: Leetcode_submissions/problems/3sum/solution.cpp
class Solution {
public:
    vector<vector<int>> ans;
    int binser(vector<int> & nums, int t, int l, int r){
        if (r >= nums.size()) return -1;
        while (r >= l){
            int m = l + (r-l)/2;
            if (nums[m] == t){
                return m;
            }else if(nums[m] > t){
                r = m-1;
            }else{
                l = m+1;
            }
        }
        return -1;
    }
    vector<vector<int>> threeSum(vector<int>& nums) {
        if (nums.size() < 3) return ans;
        
        // to prune the search space;
        vector<int> newnums;
        int prev = nums[0], count = 1;
        for (int i=1;i<nums.size();i++){
            if (nums[i] != prev){
                newnums.push_back(prev);
                if (count >= 2) newnums.push_back(prev);
                if (count >= 3) newnums.push_back(prev);
                prev = nums[i];
                count = 1;
            }else{
                count++;
            }
        }
        
        newnums.push_back(prev);
        if (count >= 2) newnums.push_back(prev);
        if (count >= 3) newnums.push_back(prev);
        
        // sort the input
        nums = newnums;
        int n = nums.size();
        sort(nums.begin(), nums.end());
        cout << n << endl;
        
        // search 3sum
        set<vector<int> > st;
        for (int i=0;i<n;i++){
            for (int j=i+1;j<n;j++){
                int res = binser(nums, 0 - (nums[i]+nums[j]), j+1, n-1);
                if (res != -1){
                    vector<int> a;
                    a.push_back(nums[i]);
                    a.push_back(nums[j]);
                    a.push_back(nums[res]);
                    if (st.find(a) != st.end()) continue;
                     st.insert(a);
                    ans.push_back(a);
                }
            }
        }
        return ans;
    }
};

File: Leetcode_submissions/problems/3sum/solution.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        n, p = [],[]
        N, P, Z = set(), set(), 0
        ans = set()
        for i in nums:
            if i == 0:
                Z += 1
            elif i > 0:
                p.append(i)
                P.add(i)
            else:
                n.append(i)
                N.add(i)
        
        if Z >= 1:
            for i in P:
                if -1 * i in N:
                    ans.add((-1*i, 0, i))
        
        if Z >= 3:
            ans.add((0,0,0))
        for i in range(len(n)):
            for j in range(i+1, len(n)):
                target = -1*(n[i]+n[j])
                if target in P:
                    ans.add(tuple(sorted([n[i], n[j], target])))
        
        for i in range(len(p)):
            for j in range(i+1, len(p)):
                target = -1*(p[i]+p[j])
                if target in N:
                    ans.add(tuple(sorted([p[i], p[j], target])))
        
        
        
        
        return list(ans)
                
                    
            
            
                
        
        